examples/FMP.jl

```diff
@@ -0,0 +1,443 @@
+# # Force Based Motion Planning
+# ![](circle_swap_pretty.gif)
+
+# This is an example implementation of the [Force Based Motion Planning (FMP)](https://arxiv.org/pdf/1909.05415.pdf) algorithm. The algorithm is a decentralized motion
+# planning algorithm where individual agents experience attractive forces towards
+# a goal position and repulsive forces from other agents or objects in the environment.
+# The FMP algorithm is included as a predefined `update_vel` function for use
+# with [`ContinuousSpace`](@ref).
+
+
+# ## Defining the Agent
+
+# First we need to define an `AbstractAgent` struct that has the necessary
+# information for the FMP algorithm to run. `pos`, and `vel` are position and
+# velocity, respectively. `tau` is an agents goal position. `color` is an
+# optional parameter for plotting used in this example. `type` is the agent
+# type: `:A` is an agent, `:O` is an obstacle, and `:T` is a target. `radius`
+# is the radius of the agent. `SSdims` is an optional parameter used for
+# plotting purposes.
+#
+# Agent types have the following distinctions:
+# - :A = agent (has velocity, subject to repulsive forces)
+# - :O = object (has no velocity, cannot move, agents must go around)
+# - :T = target (no velocity, just a projection of agent target zone)
+
+using Agents, Random, LinearAlgebra
+
+mutable struct FMP_Agent <: AbstractAgent
+    id::Int
+    pos::NTuple{2, Float64}
+    vel::NTuple{2, Float64}
+    tau::NTuple{2, Float64}
+    color::String
+    type::Symbol
+    radius::Float64
+    SSdims::NTuple{2, Float64}  ## include this for plotting
+    Ni::Array{Int64} ## array of tuples with agent ids and agent positions
+    Gi::Vector{Int64} ## array of neighboring goal IDs
+end
```

Comment:
Out of these parameters, how many are **mandatory** for the FMP algorithm? If there are more than just `pos, vel`, we could extend the `@agent` macro here.

Reply:
I will use the `@agent` macro in the example. Realistically we need `id`, `pos`, `vel`, `tau`, `type`, and `Ni`. 

`color` is really just for plotting to make the visualizations look nice. `radius` is right now used just for plotting and in my example I pull it from `model.FMP_params.d`. This is probably not a good idea because `model.FMP_params.d` is what is actually used for the FMP force computations. `SSdims` is really just used for plotting. `Gi` is a list of goal objects near an agent that I added in for my thesis that isn't arguably part of the FMP implementation but I do think it is useful so that users don't need to poll `interacting_pairs` themselves. If you want me to remove that utility I am open to it. 

I think the big question I have is how to handle plotting? Without modification, the default agent size when plotted is pretty small which is why I have a plot function like `as_f(a) = 1200*1/minimum(a.SSdims)*a.radius`. As I understand it, when plotting a simulation, the `ac, as, am` functions only pass the agent as an argument. My preference would be to call `model.space.extent` in lieu of `agent.SSdims` but I am not sure how to make that possible. If you have an idea on how to access `model` within `ac,as,am` then we could remove `SSdims` and potentially `radius` and possibly `color`.